util/collections/machine_filters.go

```diff
@@ -158,6 +158,35 @@ func HasUnhealthyCondition(machine *clusterv1.Machine) bool {
 	return conditions.IsFalse(machine, clusterv1.MachineHealthCheckSucceededCondition) && conditions.IsFalse(machine, clusterv1.MachineOwnerRemediatedCondition)
 }
 
+// HasUnhealthyControlPlaneComponentCondition returns a filter to find all unhealthy control plane machines that
+// does not have a Kubernetes node or have any of the follwing control plane component conditions set to False:
+// APIServerPodHealthy, ControllerManagerPodHealthy, SchedulerPodHealthy, EtcdPodHealthy(if using managed etcd).
+// It is different from the HasUnhealthyCondition func which checks MachineHealthCheck conditions.
+func HasUnhealthyControlPlaneComponentCondition(isEtcdManaged bool) Func {
```

Comment:
worth adding a test for this?

Reply:
the UT was added in controlplane/kubeadm/internal/controllers/scale_test.go#475

Reply:
That test is for a different function that just happens to call this one.

One could argue that you could test this one indirectly through scale_test.go, but then you need to add more cases to the scale tests to cover all the ground (there are at least 6 test cases I can see of the bat for this function). That would complicate the scale tests and add more noise there, even when the scale tests don't care about all the different cases that this function covers.

In addition, this is an exported function in a non internal package. If we were to stop using it in the scale code, then it would become untested but still part of our public api.

Reply:
Sure. We can add test for HasUnhealthyControlPlaneComponentCondition() in util/collections/machine_filters_test.go.

Reply:
done